edn/src/pp.rs

```diff
@@ -0,0 +1,240 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::borrow::Cow;
+use std::io;
+use pretty;
+
+/// A sum type representing printing instructions in a pretty print tree.
+#[derive(Debug)]
+pub enum PrintInstr<'a> {
+    /// Print a newline.
+    PrintNewline,
+    /// Print a string.
+    PrintString(Cow<'a, str>),
+    /// Print a newline, the children in the tree as a continuous block
+    /// nested at `usize`, then another newline.
+    NestChildren(usize),
+}
+
+/// A tree of pretty print instructions.
```

Comment:
The approach with having a single recursive function implementing all of the pretty printing was the first thing I did: it was sufficient and worked fine. I wanted to go a step further and build something that allows for pretty printing configuration, is type-agnostic, and abstracts away some weirdness/peculiarities of pretty.rs. A tree-like structure of instructions that can be built through macro invocations seemed like one way to go (basically, `PrettyPrintTree` is just a wrapper around pretty.rs). The advantage argument comes from ergonomics and not necessity. There are admittedly other approaches that would've been valid.

The intermediate representation also allows for easily having different pretty printing rules (e.g. for tabular data, json-like expansion etc.), that are (admittedly subjectively) easier to write in a relatively more declarative manner (which all of these macros are for). One version of this patch had more involved macros that allowed for the entire pretty printing implementation to be just a couple of lines of macro invocations in a single `From`. I've scrapped them though because they seemed a bit harder to read in the end, but maybe those would've been better to highlight my point.